source/common/websocket/codec.cc

```diff
@@ -0,0 +1,218 @@
+    frame_.opcode_ = opcode;
+    // Set final fragment flag
+    frame_.final_fragment_ = flags_and_opcode & 0x80;
+    return true;
+  }
+  decoding_error_ = true;
+  return false;
+}
+
+void Decoder::frameMaskFlag(uint8_t mask_and_length) {
+  // Set masked flag
+  if (mask_and_length & 0x80) {
+    masking_key_length_ = kMaskingKeyLength;
+  } else {
+    masking_key_length_ = 0;
+  }
+  // Set length (0 to 125) or length flag (126 or 127)
+  length_ = mask_and_length & 0x7F;
+}
+
+void Decoder::frameMaskingKey() {
+  frame_.masking_key_ = masking_key_;
+  masking_key_ = 0;
+}
+
+void Decoder::frameDataStart() {
+  frame_.payload_length_ = length_;
+  frame_.payload_ = std::make_unique<Buffer::OwnedImpl>();
+  if (length_ == 0) {
+    frameDataEnd();
+    state_ = State::KFrameHeaderFlagsAndOpcode;
+  } else {
+    state_ = State::KFramePayload;
+  }
+}
+
+void Decoder::frameData(const uint8_t* mem, uint64_t length) { frame_.payload_->add(mem, length); }
+
+void Decoder::frameDataEnd() {
+  if (!output_) {
+    output_ = std::vector<Frame>();
+  }
+  output_.value().push_back(std::move(frame_));
+  frame_.final_fragment_ = false;
+  frame_.opcode_ = 0;
+  frame_.payload_length_ = 0;
+  frame_.payload_ = nullptr;
+  frame_.masking_key_ = absl::nullopt;
+}
+
+void Decoder::doDecodeMaskFlagAndLength(const uint8_t c, const uint8_t*& mem,
+                                        uint64_t& slice_index) {
+  frameMaskFlag(c);
+  if (length_ == 0x7e) {
+    length_of_extended_length_ = kPayloadLength16Bit;
+    length_ = 0;
+    state_ = State::KFrameHeaderExtendedLength;
+  } else if (length_ == 0x7f) {
+    length_of_extended_length_ = kPayloadLength64Bit;
+    length_ = 0;
+    state_ = State::KFrameHeaderExtendedLength;
+  } else if (masking_key_length_ > 0) {
+    state_ = State::KFrameHeaderMaskingKey;
+  } else {
+    frameDataStart();
+  }
+  slice_index++;
+  mem++;
+}
+
+void Decoder::doDecodeExtendedLength(const uint8_t c, const uint8_t*& mem, uint64_t& slice_index) {
+  if (length_of_extended_length_ == 1) {
+    length_ |= static_cast<uint64_t>(c);
+  } else {
+    length_ |= static_cast<uint64_t>(c) << 8 * (length_of_extended_length_ - 1);
+  }
+  length_of_extended_length_--;
+  if (length_of_extended_length_ == 0) {
+    if (masking_key_length_ > 0) {
+      state_ = State::KFrameHeaderMaskingKey;
+    } else {
+      frameDataStart();
+    }
+  }
+  slice_index++;
+  mem++;
+}
+
+void Decoder::doDecodeMaskingKey(const uint8_t c, const uint8_t*& mem, uint64_t& slice_index) {
+  if (masking_key_length_ == 1) {
+    masking_key_ |= static_cast<uint32_t>(c);
+  } else {
+    masking_key_ |= static_cast<uint32_t>(c) << 8 * (masking_key_length_ - 1);
+  }
+  masking_key_length_--;
+  if (masking_key_length_ == 0) {
+    frameMaskingKey();
+    frameDataStart();
+  }
+  slice_index++;
+  mem++;
+}
+
+void Decoder::doDecodePayload(const Buffer::RawSlice& slice, const uint8_t*& mem,
```

Comment:
It looks like `slice` is only used to compute `remain_in_buffer`. Instead of passing in 3 different arguments how about we pass in an `absl::Span<uint8_t> data` which is basically a pointer and a length. If we did this, then `remain_in_buffer` would simply be `data.length()`

Reply:
simplified after passing `slice.len_` instead of `slice`. could to figure out a way to get `absl::Span<uint8_t>` from `RawSlice`.

Reply:
I think we can do something like this:

```
uint64_t Decoder::doDecodePayload(absl::Span<uint8_t> data) {
  if (data.length() >= length_) {
    frameData(data.data(), length_);
    state_ = State::FrameFinished;
    return length_;
  }
  frameData(data.data(), data.length());  
  return data.length();
}
```

Then when we call this method in from `inspect()` we can do:

```
  case State::FramePayload:
        uint64_t bytes_processed = doDecodePayload(absl_span<uint8_t>(mem, slice.len_ - slice_index);
        slice_index += bytes_processed;
        mem += bytes_processed;
        break;
```